consensus/src/execution_pipeline.rs

```diff
@@ -87,22 +119,41 @@ impl ExecutionPipeline {
         execute_block_tx: mpsc::UnboundedSender<ExecuteBlockCommand>,
     ) {
         while let Some(PrepareBlockCommand {
-            block_id,
+            block,
+            metadata,
             txns_to_execute,
             maybe_block_gas_limit,
             parent_block_id,
+            txn_filter,
+            txn_shuffler,
+            txn_deduper,
             result_tx,
         }) = prepare_block_rx.recv().await
         {
-            debug!("prepare_block received block {}.", block_id);
+            debug!("prepare_block received block {}.", block.id());
+            let input_txns = Self::get_shuffled_txns_to_execute(
+                &block,
+                txns_to_execute,
+                txn_filter,
+                txn_shuffler,
+                txn_deduper,
+            );
+
+            let txns_to_execute = block.transactions_to_execute_for_metadata(
+                input_txns.clone(),
```

Comment:
isn't this expensive?

Reply:
No easy way to avoid these clones.. this is a bit expensive but happening outside of critical path, so that's okay.